Description: Improve lock file handling to avoid false
 "apt-mirror is already running" messages.
Author: Ivan Borzenkov <ivan1986@list.ru>
Bug-Debian: http://bugs.debian.org/589167

--- a/apt-mirror
+++ b/apt-mirror
@@ -86,6 +86,7 @@
 use File::Copy;
 use File::Path;
 use File::Basename;
+use Fcntl qw(:flock);
 
 my $config_file;
 
@@ -178,17 +179,16 @@
 }
 
 sub lock_aptmirror {
-	system ("touch " . get_variable("var_path") . "/apt-mirror.lock");
-}
-
-sub check_lock {
-	if(-e get_variable("var_path") . "/apt-mirror.lock")
+	open (LOCK_FILE, '>', get_variable("var_path") . "/apt-mirror.lock");
+	my $lock = flock(LOCK_FILE, LOCK_EX | LOCK_NB);
+	if (!$lock)
 	{
 		die("apt-mirror is already running, exiting");
 	}
 }
 
 sub unlock_aptmirror {
+	close(LOCK_FILE);
 	unlink(get_variable("var_path") . "/apt-mirror.lock");
 }
 
@@ -296,12 +296,6 @@
 #
 #######################################################################################
 
-check_lock();
-
-$SIG{INT} = "unlock_aptmirror";
-$SIG{HUP} = "unlock_aptmirror";
-$SIG{TERM} = "unlock_aptmirror";
-
 lock_aptmirror();
 
 
